### DISCUSSION TITLE
Suggestion: separate install.cmd and run.cmd (instead of webui.cmd)

### DISCUSSION THREAD

**Discussion by WSzP** [hlky/stable-diffusion#74] [Ideas]
I had some bugs with this running the latest Anaconda on Windows 11. 

The culprit was webui.cmd, it failed to switch the environment, although it is created and visible by conda. So what I did was a one-line START.cmd INSTEAD of webui.cmd. The side effect was that the arguments to webui.py are easier to pass. So I suggest having a separate install.cmd and a run.cmd, so all the pip commands, checks, etc don't need to run again. Also, you can bypass the launcher if you are running locally and want to debug manually.

### Cowboy code START.cmd
This is my messy first try at START.cmd:
`call %windir%\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy ByPass -NoExit -Command "& 'C:\ProgramData\Anaconda3\shell\condabin\conda-hook.ps1' ; conda activate 'C:\ProgramData\Anaconda3\envs\myld' ; cd C:\local\hlky-stable-diffusion ; python scripts/webui.py --no-half --precision=full"`

You need to manually replace some string if you want to use it:
 "C:\ProgramData\Anaconda3\shell\condabin\conda-hook.ps1" with your default conda hook powershell (should be in condabin, but that can be in C:\Users\<username>\anaconda3\shell\condabin\conda-hook.ps1)
  'C:\ProgramData\Anaconda3\envs\myld' with the environment you created for this project (can be anything)
'C:\local\hlky-stable-diffusion' with the paths to where you cloned this git repo.
Hope this help.

### Example with default Anaconda3 directories
For example if you let the install script run with the default settings it should be something like:
`call %windir%\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy ByPass -NoExit -Command "& 'C:\ProgramData\Anaconda3\shell\condabin\conda-hook.ps1' ; conda activate 'C:\Users\<username>\.conda\envs\ldo' ; cd C:\local\hlky-stable-diffusion ; python scripts/webui.py --no-half --precision=full"`

### Why no relauncher? 
What the relauncher does is that if the code fails it will relaunch in 0.5 seconds. This is helpful if you run it as a remote server, but locally this makes bugfixing a bit harder, due to the CMD window being full of relaunch messages. The other problem is that if your GPU can support it, I assume you want to run it with `--no-half --precision=full` and that is easier to put it here, than overwriting the relauncher script.

**Comment by hlky** [comment]
On the first run when the environment does not exist 
`conda env create -n "%conda_env_name%" -f environment.yaml`
creates the environment, however on the next run of webui.cmd
`CondaValueError: prefix already exists: C:\Users\user\.conda\envs\ldm`
It does nothing
`conda env update -n "%conda_env_name%" --file environment.yaml --prune`
doesn't do anything unless the environment described in the environment.yaml file differs from the named environment

> so all the pip commands, checks, etc don't need to run again

So they don't actually run again, if they are for you that's something that can looked at.

I can understand where you're coming from, as you're someone with an extensive background in UX some of the design decisions might seem a bit odd, but due to the way the project started and the distribution of different guides and communities relying on this repo we've tried to 1) make things extremely simplified and 2) minimize changes to things other than the webui itself to try to avoid breaking anyone's setup. It's taken a few days to get almost everyone onto using webui.cmd which has resolved most installation issues.

I am happy to add a separate install and run but will keep the main webui.cmd as is, at least for now

**Reply by WSzP** [reply]
Thank you so much for your reply. Now I realize that my case was indeed not typical. Mainly because I use Anaconda quite extensively for a lot of projects. If it works for regular users, by all means, don't change it.
I got some comments:
1. What is the official way to run it webui.py with `--no-half --precision=full`? 
2. _"as you're someone with an extensive background in UX some of the design decisions might seem a bit odd"_ thank you, but I think that is a great start. I'm a big believer in continuous improvement (the Kaizen method). In simpler words: don't let perfect be the enemy of done. You made it happen, and that's infinitely more than what most other people did. I'm sure this can be improved further based on real user testing, gathering feedback, creating the mental model of the users, and ultimately turning it into a production-ready tool. But that thing doesn't happen overnight. And you, my friend, made something incredibly good that is available now. If you need help in the future, happy to provide it as well, just tell me.

**Reply by hlky** [reply]
[Command line options](https://github.com/hlky/stable-diffusion-webui/wiki/Command-line-options) are now documented here, documentation is something that needs to be worked on

**Reply by WSzP** [reply]
> [Command line options](https://github.com/hlky/stable-diffusion-webui/wiki/Command-line-options) are now documented here, documentation is something that needs to be worked on

I'm aware. But I thought that it might be easier to offer that as a separate cmd. First of all, I plan to NOT overwrite your code, if possible. And this is the advice I would give to most users: **don't overwrite, extend if you must**. That way when you update, I can just git pull and be happy with it, no need to worry about anything. 
But you are right, in the sense that  scripts/relauncher.py is a very simple script, doing a very simple job, so even non-developers could edit it. (But they probably won't.)
